src/MassTransit.MessagePack/Serialization/MassTransitMessagePackFormatterResolver.cs

```diff
@@ -0,0 +1,167 @@
+            new(typeof(JobStarted), typeof(InterfaceConcreteMapFormatter<JobStarted, JobStartedEvent>)),
+            new(typeof(JobStatusCheckRequested), typeof(InterfaceConcreteMapFormatter<JobStatusCheckRequested, JobStatusCheckRequestedEvent>)),
+            new(typeof(JobSubmissionAccepted), typeof(InterfaceConcreteMapFormatter<JobSubmissionAccepted, JobSubmissionAcceptedResponse>)),
+            new(typeof(JobSubmitted), typeof(InterfaceConcreteMapFormatter<JobSubmitted, JobSubmittedEvent>)),
+            new(typeof(RetryJob), typeof(InterfaceConcreteMapFormatter<RetryJob, RetryJobCommand>)),
+            new(typeof(RunJob), typeof(InterfaceConcreteMapFormatter<RunJob, RunJobCommand>)),
+            new(typeof(SaveJobState), typeof(InterfaceConcreteMapFormatter<SaveJobState, SaveJobStateCommand>)),
+            new(typeof(SetConcurrentJobLimit), typeof(InterfaceConcreteMapFormatter<SetConcurrentJobLimit, SetConcurrentJobLimitCommand>)),
+            new(typeof(SetJobProgress), typeof(InterfaceConcreteMapFormatter<SetJobProgress, SetJobProgressCommand>)),
+            new(typeof(StartJob), typeof(InterfaceConcreteMapFormatter<StartJob, StartJobCommand>)),
+            new(typeof(StartJobAttempt), typeof(InterfaceConcreteMapFormatter<StartJobAttempt, StartJobAttemptCommand>))
+        #endif
+        };
+
+        _cachedFormatters = new ConcurrentDictionary<Type, IMessagePackFormatter>();
+    }
+
+    public IMessagePackFormatter<T> GetFormatter<T>()
+    {
+        var tType = typeof(T);
+
+        if (_cachedFormatters.TryGetValue(tType, out var cachedFormatter))
+        {
+            return (IMessagePackFormatter<T>)cachedFormatter;
+        }
+
+        if (TryGetMappedType(tType, out var mappedPair))
+        {
+            var formatterType = mappedPair.Value;
+
+            EnsureFormatterTypeHasGenericParametersOfOriginalType(tType, ref formatterType);
+
+            var formatter = (IMessagePackFormatter)Activator.CreateInstance(formatterType);
+            _cachedFormatters[tType] = formatter;
+            return (IMessagePackFormatter<T>)formatter;
+        }
+
+        if (!typeof(T).IsInterface)
+        {
+            // If we have no mapper for the type, and it's not an interface, we can't create a formatter.
+            return null;
+        }
+
+        var createdConcreteFormatter = new InterfaceMessagePackFormatter<T>();
+        _cachedFormatters[tType] = createdConcreteFormatter;
+
+        return createdConcreteFormatter;
+    }
+
+    static void EnsureFormatterTypeHasGenericParametersOfOriginalType(Type tType, ref Type formatterType)
+    {
+        if (!formatterType.IsGenericTypeDefinition)
+        {
+            return;
+        }
+
+        formatterType = formatterType
+            .MakeGenericType(tType.GenericTypeArguments);
+    }
+
+    bool TryGetMappedType(Type originType, out KeyValuePair<Type, Type> mappedPair)
+    {
+        foreach (var mapPair in _mappedFormatterByType)
```

Comment:
I guess I'm curious why you didn't follow the same approach as the STJ serializer, and use a separate collection for open generic and non-generic interface message types, so that a simple dictionary lookup can be done for the non-generic mappings, since no runtime code generation is required for those types (the reason it's there).

Reply:
I have created a new PR with this optimization: #5581